process/rating/blockSigningRater.go

```diff
@@ -0,0 +1,106 @@
+package rating
+
+import (
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/economics"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+// BlockSigningRater defines the behaviour of a struct able to do ratings for validators
+type BlockSigningRater struct {
+	sharding.RatingReader
+	startRating                 uint32
+	maxRating                   uint32
+	minRating                   uint32
+	proposerIncreaseRatingStep  int32
+	proposerDecreaseRatingStep  int32
+	validatorIncreaseRatingStep int32
+	validatorDecreaseRatingStep int32
+}
+
+//NewBlockSigningRater creates a new RaterHandler of Type BlockSigningRater
+func NewBlockSigningRater(ratingsData *economics.RatingsData) (*BlockSigningRater, error) {
+	if ratingsData.MinRating() > ratingsData.MaxRating() {
+		return nil, process.ErrMaxRatingIsSmallerThanMinRating
+	}
+	if ratingsData.MaxRating() < ratingsData.StartRating() || ratingsData.MinRating() > ratingsData.StartRating() {
+		return nil, process.ErrStartRatingNotBetweenMinAndMax
+	}
+
+	return &BlockSigningRater{
+		startRating:                 ratingsData.StartRating(),
+		minRating:                   ratingsData.MinRating(),
+		maxRating:                   ratingsData.MaxRating(),
+		proposerIncreaseRatingStep:  int32(ratingsData.ProposerIncreaseRatingStep()),
+		proposerDecreaseRatingStep:  int32(0 - ratingsData.ProposerDecreaseRatingStep()),
+		validatorIncreaseRatingStep: int32(ratingsData.ValidatorIncreaseRatingStep()),
+		validatorDecreaseRatingStep: int32(0 - ratingsData.ValidatorDecreaseRatingStep()),
+	}, nil
+}
+
+func (bsr *BlockSigningRater) computeRating(ratingStep int32, val uint32) uint32 {
+	newVal := int64(val) + int64(ratingStep)
+	if newVal < int64(bsr.minRating) {
+		return bsr.minRating
+	}
+	if newVal > int64(bsr.maxRating) {
+		return bsr.maxRating
+	}
+
+	return uint32(newVal)
+}
+
+//GetRating returns the Rating for the specified public key
+func (bsr *BlockSigningRater) GetRating(pk string) uint32 {
+	if bsr.RatingReader == nil {
+		return 1
+	}
+	return bsr.RatingReader.GetRating(pk)
+}
+
+//GetRatings gets all the ratings that the current rater has
+func (bsr *BlockSigningRater) GetRatings(addresses []string) map[string]uint32 {
+	if bsr.RatingReader == nil {
```

Comment:
shouldn't this always exist ? if it can be nil sometimes, than you have to check with check.ifNil

Reply:
fixed with the above NilRatingReader